bam_stat.c

```diff
@@ -97,15 +97,130 @@ static void usage_exit(FILE *fp, int exit_status)
 {
     fprintf(fp, "Usage: samtools flagstat [options] <in.bam>\n");
     sam_global_opt_help(fp, "-.---@");
+    fprintf(fp, "  -O, --");
+    fprintf(fp, "output-fmt FORMAT[,OPT[=VAL]]...\n"
+            "               Specify output format (json, tsv)\n");
     exit(exit_status);
 }
 
+
+static void out_fmt_default(bam_flagstat_t *s)
+{
+    char b0[16], b1[16];
+    printf("%lld + %lld in total (QC-passed reads + QC-failed reads)\n", s->n_reads[0], s->n_reads[1]);
+    printf("%lld + %lld secondary\n", s->n_secondary[0], s->n_secondary[1]);
+    printf("%lld + %lld supplementary\n", s->n_supp[0], s->n_supp[1]);
+    printf("%lld + %lld duplicates\n", s->n_dup[0], s->n_dup[1]);
+    printf("%lld + %lld mapped (%s : %s)\n", s->n_mapped[0], s->n_mapped[1], percent(b0, s->n_mapped[0], s->n_reads[0]), percent(b1, s->n_mapped[1], s->n_reads[1]));
+    printf("%lld + %lld paired in sequencing\n", s->n_pair_all[0], s->n_pair_all[1]);
+    printf("%lld + %lld read1\n", s->n_read1[0], s->n_read1[1]);
+    printf("%lld + %lld read2\n", s->n_read2[0], s->n_read2[1]);
+    printf("%lld + %lld properly paired (%s : %s)\n", s->n_pair_good[0], s->n_pair_good[1], percent(b0, s->n_pair_good[0], s->n_pair_all[0]), percent(b1, s->n_pair_good[1], s->n_pair_all[1]));
+    printf("%lld + %lld with itself and mate mapped\n", s->n_pair_map[0], s->n_pair_map[1]);
+    printf("%lld + %lld singletons (%s : %s)\n", s->n_sgltn[0], s->n_sgltn[1], percent(b0, s->n_sgltn[0], s->n_pair_all[0]), percent(b1, s->n_sgltn[1], s->n_pair_all[1]));
+    printf("%lld + %lld with mate mapped to a different chr\n", s->n_diffchr[0], s->n_diffchr[1]);
+    printf("%lld + %lld with mate mapped to a different chr (mapQ>=5)\n", s->n_diffhigh[0], s->n_diffhigh[1]);
+}
+
+static void out_fmt_json(bam_flagstat_t *s) {
+    char b0[16], b1[16];
+    char *str = "{ \"QC-passed reads\": {\
```

Comment:
Is there any reason why you chose to use a separate variable for the format string (apart from tidiness)?

It's generally a bad idea to do this as it stops the compiler from checking that the format string matches the types of the parameters passed in.  For example, trying to compile:

```c
#include <stdio.h>
int main(int argc, char **argv) {
    long long int answer = 42;
    const char *fmt = "The answer is %d\n";

    printf(fmt, answer);

    printf("The answer is %d\n", answer);
}
```

results in:
```
$ cc test.c
test.c: In function ‘main’:
test.c:8:28: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long long int’ [-Wformat=]
     printf("The answer is %d\n", answer);
                           ~^
                           %lld

```

So it accepted the first `printf` but not the second, even though both have the same bug.

Also, the spaces due to the indentation following each backslashed newline is ending up in the JSON output.  The best way to do multi-line strings in C is to use the fact that `"AB" "CD"` is equivalent to `"ABCD"` and put each line in its own set of double-quotes, ending up with something like this:
```c
printf("{ \"QC-passed reads\": {"
       " \"total\": \"%lld\","
       // Bit in middle skipped
       " \"with mate mapped to a different chr (mapQ >= 5)\": \"%lld\" } "
       "}\n",
       s->n_reads[0],
       // ... variables skipped ...
       s->n_diffhigh[1]);
```

You might also want to consider putting a `\n` on the end of each line so you get a pretty-printed JSON file.  This would both make it easier to read and allow simple tools like `awk` or `grep` to easily extract individual lines.  Any reasonable JSON parser should be able to handle the extra white-space.

Reply:
1 -- Your intuition is correct. I mostly used for tidiness, but your idea is valid. I revised the code.
2 -- Agreed. I updated the code.